docs/book/rest-api.md

```diff
@@ -2211,3 +2211,71 @@ The `expressions` field is an array of the results of evaluating each of the exp
 The `bindings` field is a JSON object mapping `string`s to JSON values that describe the variable bindings resulting from evaluating the query.
 
 If the watch was set on a data reference instead of a query, the `result` field will simply be the value of the document requested, instead of an array of values.
+
+## <a name="diagnostics"></a> Diagnostics
+
+The OPA server has the capability to log and return diagnostics on past requests to the user. By default, the server will not log any diagnostics at all. In order to have the server start logging diagnostic information, the document located at `data.system.diagnostics.config` (henceforth referred to as the config) must be defined. The config must evaluate to an object, which must be structured as follows:
+
+```
+{
+    "mode": <"all"/"none">,
+    "metrics": <true/false>,
+    "explain": <true/false>
+}
+```
+
+Whenever a data GET/POST or query GET request is received by the server, it evaluates the config document to determine whether or not diagnostics should be logged. If `config.mode` is "all", then all diagnostics are turned on for the request. Similarly, if `config.mode` is "none", then no diagnostics are recorded for the request. If `config.mode` has neither of those values, the value of `config.metrics` and `config.explain` are checked to determine if metrics and explanations should be recorded for the request, respectively.
+
+If the config document does not conform to the structure above, then diagnostics are automatically disabled.
+
+In order to allow the config document to make dynamic decisions about whether or not to record diagnostics for a given request, the server will provide a special input document when evaluating it. The input document will contain information from the HTTP request that was issued to the server, of the form:
+
+```
+{
+    "method": <HTTP request method>,
+    "path": <Full HTTP request path>,
+    "body": <HTTP request body (null if not present)>,
+
+    # The query parameters in the HTTP request.
+    "params": {
+        "param0": ["value0", "value1", ...],
+        "param1": [...],
+        ...
+    },
+
+    # The header fields in the HTTP request.
+    "header": {
+        "field0": ["value0", "value1", ...],
+        "field1": [...],
+        ...
+    }
+}
+```
+
+Diagnostics may be fetched from the server using the Data GET endpoint. When the server sees a GET request to `/v1/data/system/diagnostics`, it will not evaluate the document located there, but instead return a list of the diagnostics stored on the server (ordered from oldest to newest). The server will honor the `explainMode` and `pretty` paramters from the GET request, but the others will be ignored. The response is of the form:
+
+```
+[
+    {
+        "id": <sequence number>,
```

Comment:
I see that this ID is auto-generated by the diagnostics package. I'd rather see this ID generated by the server (e.g., we currently have a request ID generated by the request logger) and propagated throughout. This probably requires a bit of work that's outside the scope of this PR so for the time being I'd remove this.